File: resources/engine/switchmap.py
# -*- coding: utf-8 -*-
"""
javac компилирует `switch (enumVar) { case CONST: ... }` не напрямую, а через
синтетический вложенный класс вида Outer$1 с полем
`static final int[] $SwitchMap$pkg$EnumType`, которое заполняется в
<clinit> через `array[EnumType.CONST.ordinal()] = N;` (в try/catch
NoSuchFieldError на каждую константу). Сам метод при этом делает
`array[enumVar.ordinal()]` и свитчится по числу N.

Этот модуль находит такие синтетические классы, вручную разбирает их
<clinit> (без полной символической интерпретации - тут это не нужно,
паттерн линейный и жёстко фиксирован) и строит обратную таблицу
N -> имя константы, чтобы движок мог восстановить настоящий
`switch (enumVar) { case CONST: ... }` вместо уродливого индекса по массиву.
"""
from ir import decode_method


def detect_switchmaps(class_files):
    """Возвращает (switchmap_fields, synthetic_class_internals):
    switchmap_fields: dict (owner_class_internal, field_name) -> {
        'enum_owner': enum_class_internal, 'table': {int_value: const_name}
    }
    synthetic_class_internals: set имён классов, которые целиком являются
    этим сгенерированным компилятором артефактом и не должны попадать в
    итоговый вывод .java (в исходнике их никогда не было)."""
    switchmap_fields = {}
    synthetic_classes = set()

    for internal, cf in class_files.items():
        candidate_fields = [f for f in cf.fields
                             if f.descriptor == "[I" and f.name.startswith("$SwitchMap$")]
        if not candidate_fields:
            continue
        clinit = next((m for m in cf.methods if m.name == "<clinit>"), None)
        if clinit is None or clinit.code is None:
            continue
        try:
            instrs, order = decode_method(clinit.code)
        except Exception:
            continue

        for f in candidate_fields:
            table = _extract_table(cf, instrs, order, f.name)
            if table:
                switchmap_fields[(internal, f.name)] = {
                    "enum_owner": table["enum_owner"],
                    "table": table["table"],
                }
        # Класс целиком - синтетический switch-map холдер (не часть исходника),
        # если ВСЕ его поля - это найденные $SwitchMap$ массивы, и единственный
        # метод - <clinit>.
        if len(cf.fields) == len(candidate_fields) and \
                all((internal, f.name) in switchmap_fields for f in candidate_fields) and \
                len([m for m in cf.methods if m.name != "<clinit>"]) == 0:
            synthetic_classes.add(internal)

    return switchmap_fields, synthetic_classes


def _extract_table(cf, instrs, order, field_name):
    """Линейно ищем паттерн:
        getstatic  <field_name>:[I
        getstatic  EnumOwner.CONST:LEnumOwner;
        invokevirtual  ...ordinal:()I
        <push int N>            (bipush/sipush/iconst_N)
        iastore
    Границы try/catch(NoSuchFieldError) не мешают - сканируем сырую
    последовательность инструкций, игнорируя exception table целиком."""
    table = {}
    enum_owner = None
    seq = [instrs[pc] for pc in order]
    n = len(seq)
    i = 0
    while i < n:
        ins = seq[i]
        if ins.mnemonic == "getstatic":
            r = cf.ref_string(ins.cp_index)
            if r and r[1] == field_name and r[2] == "[I":
                # ищем непосредственно следующие 4 инструкции
                if i + 4 < n:
                    g2, ov, push_ins, store = seq[i + 1], seq[i + 2], seq[i + 3], seq[i + 4]
                    if g2.mnemonic == "getstatic" and ov.mnemonic == "invokevirtual" and \
                            store.mnemonic == "iastore":
                        r2 = cf.ref_string(g2.cp_index)
                        rv = cf.ref_string(ov.cp_index)
                        if r2 and rv and rv[1] == "ordinal":
                            owner = r2[0]
                            const_name = r2[1]
                            val = _push_int_value(push_ins)
                            if val is not None:
                                table[val] = const_name
                                enum_owner = owner
                                i += 5
                                continue
        i += 1
    if not table or enum_owner is None:
        return None
    return {"enum_owner": enum_owner, "table": table}


def _push_int_value(ins):
    if ins.mnemonic.startswith("iconst_"):
        v = ins.mnemonic.split("_")[1]
        return -1 if v == "m1" else int(v)
    if ins.mnemonic in ("bipush", "sipush"):
        return ins.ival
    return None

File: resources/engine_cpp/src/switchmap.cpp
// switchmap.cpp - см. switchmap.hpp. 1:1 порт switchmap.py.
#include "switchmap.hpp"

#include <algorithm>
#include <optional>

#include "ir.hpp"

namespace nd {

namespace {

std::optional<int64_t> push_int_value(const Instruction& ins) {
    if (ins.mnemonic.rfind("iconst_", 0) == 0) {
        std::string v = ins.mnemonic.substr(7);
        if (v == "m1") return -1;
        return std::stoll(v);
    }
    if (ins.mnemonic == "bipush" || ins.mnemonic == "sipush") return ins.ival;
    return std::nullopt;
}

std::optional<SwitchmapFieldInfo> extract_table(const ClassFile& cf, const DecodedMethod& dm, const std::string& field_name) {
    std::map<int64_t, std::string> table;
    std::optional<std::string> enum_owner;

    std::vector<const Instruction*> seq;
    seq.reserve(dm.order.size());
    for (auto pc : dm.order) seq.push_back(&dm.instrs.at(pc));
    size_t n = seq.size();

    size_t i = 0;
    while (i < n) {
        const Instruction& ins = *seq[i];
        if (ins.mnemonic == "getstatic" && ins.cp_index.has_value()) {
            auto r = cf.ref_string(static_cast<uint16_t>(*ins.cp_index));
            if (r.has_value() && std::get<1>(*r) == field_name && std::get<2>(*r) == "[I") {
                if (i + 4 < n) {
                    const Instruction& g2 = *seq[i + 1];
                    const Instruction& ov = *seq[i + 2];
                    const Instruction& push_ins = *seq[i + 3];
                    const Instruction& store = *seq[i + 4];
                    if (g2.mnemonic == "getstatic" && ov.mnemonic == "invokevirtual" && store.mnemonic == "iastore") {
                        auto r2 = g2.cp_index.has_value() ? cf.ref_string(static_cast<uint16_t>(*g2.cp_index)) : std::nullopt;
                        auto rv = ov.cp_index.has_value() ? cf.ref_string(static_cast<uint16_t>(*ov.cp_index)) : std::nullopt;
                        if (r2.has_value() && rv.has_value() && std::get<1>(*rv) == "ordinal") {
                            std::string owner = std::get<0>(*r2);
                            std::string const_name = std::get<1>(*r2);
                            auto val = push_int_value(push_ins);
                            if (val.has_value()) {
                                table[*val] = const_name;
                                enum_owner = owner;
                                i += 5;
                                continue;
                            }
                        }
                    }
                }
            }
        }
        i += 1;
    }

    if (table.empty() || !enum_owner.has_value()) return std::nullopt;
    return SwitchmapFieldInfo{*enum_owner, table};
}

}  // namespace

SwitchmapDetectionResult detect_switchmaps(const std::map<std::string, ClassFile>& class_files) {
    SwitchmapDetectionResult result;

    for (auto& [internal, cf] : class_files) {
        std::vector<const Field*> candidate_fields;
        for (auto& f : cf.fields)
            if (f.descriptor == "[I" && f.name.rfind("$SwitchMap$", 0) == 0) candidate_fields.push_back(&f);
        if (candidate_fields.empty()) continue;

        const Method* clinit = nullptr;
        for (auto& m : cf.methods)
            if (m.name == "<clinit>") {
                clinit = &m;
                break;
            }
        if (clinit == nullptr || !clinit->has_code) continue;

        DecodedMethod dm;
        try {
            dm = decode_method(clinit->code);
        } catch (const std::exception&) {
            continue;
        }

        for (auto f : candidate_fields) {
            auto table = extract_table(cf, dm, f->name);
            if (table.has_value()) result.switchmap_fields[{internal, f->name}] = *table;
        }

        // Класс целиком - синтетический switch-map холдер (не часть
        // исходника), если ВСЕ его поля - это найденные $SwitchMap$ массивы,
        // и единственный метод - <clinit>.
        bool all_fields_are_switchmaps =
            cf.fields.size() == candidate_fields.size() &&
            std::all_of(candidate_fields.begin(), candidate_fields.end(),
                        [&](const Field* f) { return result.switchmap_fields.count({internal, f->name}) != 0; });
        size_t non_clinit_methods = 0;
        for (auto& m : cf.methods)
            if (m.name != "<clinit>") non_clinit_methods++;
        if (all_fields_are_switchmaps && non_clinit_methods == 0) result.synthetic_classes.insert(internal);
    }

    return result;
}

}  // namespace nd
